### PR TITLE
Feature/amigo free mode restart

### PR DESCRIPTION
**Pull request by KevinDang74**
Tested on the robot and it works without problems.

### CODE REVIEW

File: src/node_alive/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from autostarter import AutoStarter
```
Comment (jlunenburg) [inline src/node_alive/__init__.py:1]: "No newline at end of file"

File: src/node_alive/autostarter.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+# ROS
+import os
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: os is not a ros import

File: src/node_alive/autostarter.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+# TU/e Robotics
+from node_alive.srv import AutoStarterCommand, AutoStarterCommandRequest, AutoStarterCommandResponse
+
+class AutoStarter(object):
+    """
+    Launches a launch file and waits for a command to either shutdown or to restart.
+    It is also possible to start other launch files that are stored on the parameter server, like the demo's.
+    """
+    launch_files_list = []
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: Why is this 'class variable' launch_files_list defined here?

File: src/node_alive/autostarter.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import rospy
+from std_msgs.msg import String
+
+# TU/e Robotics
+from node_alive.srv import AutoStarterCommand, AutoStarterCommandRequest, AutoStarterCommandResponse
+
+class AutoStarter(object):
+    """
+    Launches a launch file and waits for a command to either shutdown or to restart.
+    It is also possible to start other launch files that are stored on the parameter server, like the demo's.
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: The docstring might not be entirely clear to people who're not involved in your project. What are "launch files that are stored on the parameter server" and what does "like the demo's" mean?

File: src/node_alive/autostarter.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    """
+    launch_files_list = []
+
+    def __init__(self):
+        """
+        Loads the launch files that can be started from the parameter server.
+        Creates a service over which the command will be received, a publisher that publishes the current launch file.
+        Set some variables.
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: Idem. "Set some variables" also doesn't add anything

File: src/node_alive/autostarter.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        self.launch_files_list = rospy.get_param('~launch_files_list')
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: Is there a reason why certain variables are private and some public (with or without underscore)?

File: src/node_alive/autostarter.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        Recieves the request from the service and sets it to pending. As long as there is no response, the function
+        will sleep till it gets one. After that the response will reset.
+        :param req: the command request that the service receives from the client contains two data types.
+        :return:
+        """
+        self._pending_resp = None
+        self._pending_req = req
+        rate = rospy.Rate(10.0)
+        while self._pending_resp is None and not rospy.is_shutdown():
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: Maybe this should set pending resp back to None. I.e.,
```
response = self._pending_resp
self._pending_resp = None
return response
```

File: src/node_alive/autostarter.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            rate.sleep()
+        return self._pending_resp
+
+    def _handle_auto_starter_command(self, req):
+        """
+        All request from the service are passed into this function and depending on the value it will either shutdown,
+        restart the launch file or start another launch file.
+        :param req: the command request that the service receives from the client contains two data types.
+        :return: returns the integer 1  when function finishes successful or -1 or -2 when no or incorrect file is given.
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: Actually, an AutoStarterCommandResponse is returned. The param req is hence an AutoStarterCommandRequest

File: src/node_alive/autostarter.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        Loads the launch files that can be started from the parameter server.
+        Creates a service over which the command will be received, a publisher that publishes the current launch file.
+        Set some variables.
+        """
+        self.launch_files_list = rospy.get_param('~launch_files_list')
+        self._command_service = rospy.Service('auto_starter_command', AutoStarterCommand, self._srv_callback)
+        self._status_pub = rospy.Publisher('current_launch_file', String, queue_size=1, latch=True)
+
+        self.launch = None
```
Comment (jlunenburg) [inline src/node_alive/autostarter.py]: The member attribute name "self.launch" is not really clear. You might want to add a few comments as well.

**Review by jlunenburg — CHANGES_REQUESTED** [review]
Nice that it's finally working! I added a few comments concerning docstrings, style and clarity (for future uses). After this is merged, we can/should replace the old method.

File: src/node_alive/__init__.py (hunk 1 above)
Comment (KevinDang74) [reply src/node_alive/__init__.py:1]: added line updated in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 7 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: Updated in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 2 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: replaced it in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 3 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: removed it in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 4 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: rewritten it in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 5 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: removed in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 6 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: Changed it to variables to private and kept the necessary ones public in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 8 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: Changed in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py (hunk 9 above)
Comment (KevinDang74) [reply src/node_alive/autostarter.py]: Wrote extra comment for self.launch in a1ef39d19ac8416def3517506f5d571fcb9e1645

File: src/node_alive/autostarter.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+    def _srv_callback(self, req):
+        """
+        Recieves the request from the service and sets it to pending. As long as there is no response, the function
+        will sleep till it gets one. After that the response will reset.
+        :param req: (AutoStarterCommandResponse) the command request that the service receives from the client contains
+        two data types, which launch file and which command.
+        :return response:
+        """
+        self._time_out = 10.0
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:54]: These parameter should be set during initialization.

File: src/node_alive/autostarter.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        self._pending_resp = None
+        self._pending_req = req
+        rate = rospy.Rate(10.0)
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:57]: The rate should be parameter, which is set during initialization.

File: src/node_alive/autostarter.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        t_end = rospy.Time.now() + rospy.Duration(self._time_out)
+
+        while self._pending_resp is None and not rospy.is_shutdown():
+            rate.sleep()
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:61]: This should be placed at the end of a loop, to secure the the loop rate.

File: src/node_alive/autostarter.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+        if req.filename in self._launch_files_list:
+            if req.command == AutoStarterCommandRequest.START:
+                if self._launch is None:
+                    self.start(self._bringup_package_path , os.path.join("launch", "state_machines", req.filename))
+                    rospy.loginfo("Performed new start")
+                    return AutoStarterCommandResponse(AutoStarterCommandResponse.SUCCEEDED)
+                else:
+                    self.stop()
+                    self.start( self._bringup_package_path, os.path.join("launch", "state_machines", req.filename))
+                    rospy.loginfo("Performed restart")
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:92]: include information about launch file

File: src/node_alive/autostarter.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        restart the launch file or start another launch file.
+        :param req: (AutoStarterCommandResponse) the command request that the service receives from the client contains
+        two data types, which launch file and which command.
+        :return AutoStarterCommandResponse:
+        """
+        if req.filename in self._launch_files_list:
+            if req.command == AutoStarterCommandRequest.START:
+                if self._launch is None:
+                    self.start(self._bringup_package_path , os.path.join("launch", "state_machines", req.filename))
+                    rospy.loginfo("Performed new start")
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:87]: Include information about launch file

File: src/node_alive/autostarter.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+                    return AutoStarterCommandResponse(AutoStarterCommandResponse.SUCCEEDED)
+                else:
+                    self.stop()
+                    self.start( self._bringup_package_path, os.path.join("launch", "state_machines", req.filename))
+                    rospy.loginfo("Performed restart")
+                    return AutoStarterCommandResponse(AutoStarterCommandResponse.SUCCEEDED)
+            elif req.command == AutoStarterCommandRequest.STOP:
+                rospy.loginfo("Performing shutdown")
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:95]: Add information about launch file

File: src/node_alive/autostarter.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+                self.stop()
+                rospy.loginfo("Performed shutdown")
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:97]: Too much logging here. Maybe move this some logging to the seperate functions

File: src/node_alive/autostarter.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+    def start(self, package, path):
+        """
+        Before a launche file is started, it kills the node state_machine to make sure that all running launch files
+        are closed. Then it finds the right directory of the launch file and launches the file.
+        :param package: package of the current launch file
+        :param path: path to the current launch file
+        """
+        rosnode.kill_nodes(['/state_machine'])
+        rospy.sleep(0.1)
```
Comment (MatthijsBurgh) [inline src/node_alive/autostarter.py:112]: Does accept rospy.sleep accept floats? I doubt it. If you don't want to wait for whole seconds, you probably need to use a different argument.

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]

**Comment by MatthijsBurgh** [comment]
@KevinDang74 Please apply the requested changes or close this PR.

**Pull request closed without merging**